File: core/src/main/java/dev/jason/project/spring/vc_server/core/model/Device.java
package dev.jason.project.spring.vc_server.core.model;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.Setter;

import java.time.LocalDateTime;
import java.util.Objects;

@AllArgsConstructor
@Getter
public final class Device {
    private final String ownerUid;
    private final String name;
    private final Type type;
    private final OS os;
    private final String version;
    @Setter
    private String token;
    private final LocalDateTime lastUsed;

    public enum OS {
        Android // currently only android is supported. may add support to other oses in the future.
    }

    public enum Type {
        Mobile, Tablet // currently only mobile and tablet are supported. may add support to other devices in the future.
    }

    @Override
    public int hashCode() {
        return Objects.hash(name, os, version, token, lastUsed);
    }

    @Override
    public String toString() {
        return "Device[" +
            "name=" + name + ", " +
            "type=" + type + ", " +
            "os=" + os + ", " +
            "version=" + version + ", " +
            "lastUsed=" + lastUsed + ']';
    }

    @Override
    public boolean equals(Object other) {
        if (!(other instanceof Device)) return false;
        return Objects.equals(((Device) other).name, name) &&
            Objects.equals(((Device) other).token, token); // returns true if fcm token and name are same
    }

}

File: core/src/main/java/dev/jason/app/compose/vaultchat/core/model/Device.kt
package dev.jason.app.compose.vaultchat.core.model

import android.content.Context
import android.content.res.Configuration
import android.os.Build

data class Device(
    val ownerUid: String,
    val name: String,
    val type: Type,
    val os: Os,
    val version: String,
    val token: String,
) {

    override fun equals(other: Any?): Boolean {
        if (other !is Device) return false
        return other.name == name &&
                other.type == type &&
                other.os == os &&
                other.version == version
    }

    override fun hashCode(): Int {
        var result = name.hashCode()
        result = 31 * result + type.hashCode()
        result = 31 * result + os.hashCode()
        result = 31 * result + version.hashCode()
        result = 31 * result + token.hashCode()
        return result
    }

    enum class Os {
        Android // currently only android is supported
    }

    enum class Type {
        Mobile, Tablet
    }

    companion object {
        fun getCurrentDevice(context: Context, ownerUid: String, token: String): Device {
            return Device(
                ownerUid = ownerUid,
                name = "${Build.MANUFACTURER} ${Build.MODEL}",
                os = Os.Android,
                version = "${Build.VERSION.RELEASE}",
                token = token,
                type = getDeviceType(context)
            )
        }

        fun getDeviceType(context: Context): Type {
            val screenLayout = context.resources.configuration.screenLayout
            val mask = Configuration.SCREENLAYOUT_SIZE_MASK
            val isTablet = (screenLayout and mask) >= Configuration.SCREENLAYOUT_SIZE_LARGE
            return if (isTablet) {
                Type.Tablet
            } else {
                Type.Mobile
            }
        }
    }
}